### ISSUE TITLE
Initiating Search from a table that doesn't use an integer ID field causes search value to be coerced to NULL

### ISSUE THREAD

**Issue by ozzyaaron** [activerecord-hackery/ransack#744]
## Overview

The issue is if you start your query from a table that doesn't have the same *type* of ID column then you can end up with an incorrect final search value. i.e. starting from a model that uses a UUID ID column and searching out to an association with an Integer ID column won't work as the UUID coercer turns the Integer into a NULL.

### Library Versions etc

1. Rails 5, ActiveRecord 5 (this bug came when upgrading from Rails 4.2)
1. Ransack 1.8.2 (latest at the time)
1. Postgres 9.5.5 though this is unimportant

### Schema

```
  create_table "no_integer_ids", id: :uuid, default: -> { "uuid_generate_v4()" }, force: :cascade do |t|
    t.integer  "deprecated_sequential_id"
    t.integer  "some_association_id",                                              null: false
    ....

  create_table "somethings", force: :cascade do |t|
    t.integer  "no_integer_id_id"

   ....
```

### Models

```
class NoIntegerId
  has_many :somethings
end

class Something
  belongs_to :no_integer_id
end
```

### Output

```
NotIntegerId.ransack(somethings_id_eq: 5).result.to_sql
```

You'd expect this to output something like:

```
SELECT no_integer_ids.* FROM no_integer_ids
LEFT OUTTER JOIN somethings ON ....
WHERE somethings.id = 5;
```

Instead you get ...

```
SELECT no_integer_ids.* FROM no_integer_ids 
LEFT OUTTER JOIN somethings ON .... 
WHERE somethings.id = NULL;
```

Completely not expected 😸 

## Reason

I'm fairly certain the culprit is that the typecaster for arel_predicates use the initiator (i.e. NoIntegerId above) as the typecaster rather than the model that is eventually being queried (i.e. Something above) so if it has no id column OR the id column is a UUID then it cannot be coerced.

https://github.com/activerecord-hackery/ransack/blob/master/lib/ransack/adapters/active_record/ransack/context.rb#L42

I'm just lodging this here as I'm not even sure where to begin with this. It seems that the typecaster would need to be found and set as `Visitor` is passing through nodes. 

That being said it took a while to track down so hopefully this might help someone else.

**Comment by ozzyaaron** [comment]
We've been converting any of these we've found to custom ransackers. Is the appropriate 'fix' here to document in the wiki? If so I don't at all mind making that PR.

**Comment by seanfcarroll** [comment]
That would be really great ... thank you.

**Comment by ozzyaaron** [comment]
Cool, things are a bit busy now, but it's in an open tab so I'll get to it soon 😄

**Comment by ozzyaaron** [comment]
There are two cases that we've broadly run into and I wanted to see if these are solutions you'd be happy to promote, it's also an opportunity to see if there is something better out there. 

In the case where it's a column on the table itself then we just define a ransacker. The only case I believe here is the primary key of the table itself. If you have a model `Job` that uses a primary key column `id` of type `uuid` then a `CAST` in the ransacker (RDMS dependent).

For joining queries we've tended to write a database view in these situations that will cast for us. This seems heavyweight but it has worked pretty well for us and generally speaking allows us to roll up a lot of things that ransack is reaching out for in more efficient ways in the database view. Generally to do this with a ransack we've found it requires the use of a subquery which in many situations is just far too slow. 

As a concrete example if that `Job` model above has one `User` then generally we might query with `user_id_eq` or something. In that case we've found that we could have a DB view include the `user_id` really easily OR write a ransacker that queries the `users` table using a `WHERE` condition from the `jobs` table. We've preferred to write database views in almost all circumstances. 

If I were to write the wiki page I'd probably be outlining the 2 scenarios and pushing database views for the second scenario outlined if that is okay?
